server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -226,12 +228,10 @@ protected void doRun() throws Exception {
                 removed = unknownSnapshotShards.remove(snapshotShard);
                 assert removed : "snapshot shard to remove does not exist " + snapshotShardSize;
                 if (isMaster) {
-                    final ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes = ImmutableOpenMap.builder(
-                        knownSnapshotShards
-                    );
+                    final Map<SnapshotShard, Long> newSnapshotShardSizes = new HashMap<>(knownSnapshotShards);
                     updated = newSnapshotShardSizes.put(snapshotShard, snapshotShardSize) == null;
                     assert updated : "snapshot shard size already exists for " + snapshotShard;
-                    knownSnapshotShards = newSnapshotShardSizes.build();
+                    knownSnapshotShards = Map.copyOf(newSnapshotShardSizes);
```

Comment:
I believe Collections.unmodifiableMap might be a better choice here, since we don't modify newSnapshotShardSizes after. We shouldn't need to make a copy I think.